mlflow/server/js/src/components/MetricsPlotPanel.js

```diff
@@ -0,0 +1,183 @@
+import React from 'react';
+import { connect } from 'react-redux';
+import Utils from '../utils/Utils';
+import RequestStateWrapper from './RequestStateWrapper';
+import { getMetricHistoryApi, getUUID } from '../Actions';
+import PropTypes from 'prop-types';
+import _ from 'lodash';
+import { MetricsPlotView } from './MetricsPlotView';
+import { getRunTags } from '../reducers/Reducers';
+import { MetricsPlotControls, X_AXIS_RELATIVE, X_AXIS_STEP } from './MetricsPlotControls';
+import qs from 'qs';
+import { withRouter } from 'react-router-dom';
+import Routes from '../Routes';
+
+export const CHART_TYPE_LINE = 'line';
+export const CHART_TYPE_BAR = 'bar';
+
+class MetricsPlotPanel extends React.Component {
+  static propTypes = {
+    runUuids: PropTypes.arrayOf(String).isRequired,
+    metricKey: PropTypes.string.isRequired,
+    latestMetricsByRunUuid: PropTypes.object.isRequired,
+    distinctMetricKeys: PropTypes.arrayOf(String).isRequired,
+    metricsWithRunInfoAndHistory: PropTypes.arrayOf(Object).isRequired,
+    getMetricHistoryApi: PropTypes.func.isRequired,
+    location: PropTypes.object.isRequired,
+    history: PropTypes.object.isRequired,
+  };
+
+  constructor(props) {
+    super(props);
+    const plotMetricKeys = MetricsPlotPanel.getPlotMetricKeysFromUrl(props.location.search);
+    const selectedMetricKeys = plotMetricKeys.length ? plotMetricKeys : [props.metricKey];
+    this.state = {
+      selectedXAxis: X_AXIS_RELATIVE,
+      selectedMetricKeys,
+      showDot: false,
+      historyRequestIds: [],
+    };
+    this.loadMetricHistory(this.props.runUuids, selectedMetricKeys);
+  }
+
+  static predictChartType(metrics) {
+    // Show bar chart when every metric has exactly 1 metric history
+    if (metrics && metrics.length && _.every(metrics, (m) => m.history && m.history.length === 1)) {
+      return CHART_TYPE_BAR;
+    }
+    return CHART_TYPE_LINE;
+  }
+
+  static getPlotMetricKeysFromUrl = (search) => JSON.parse(qs.parse(search)['plot_metric_keys']);
```

Comment:
Could `qs.parse(search)['plot_metric_keys']` ever be undefined? I guess not for now, but maybe we should be defensive here? Mainly because `JSON.parse(undefined)` will throw